subiquity/models/tests/test_filesystem.py

```diff
@@ -78,3 +159,702 @@ def test_errors(self):
+        self.assertTrue(self.fsm.any_configuration_done())
+
+    def test_bootable(self):
+        """bootable True only if FSM has partition with boot/bios_grub flag"""
+        self.fsm.probe()
+        all_disks = self.fsm.all_disks()
+        disk = random.choice(all_disks)
+        partition1 = self.fsm.add_partition(disk=disk, size=(1 * MB))
+        fs1 = self.fsm.add_filesystem(partition1, 'ext4')
+        self.fsm.add_mount(fs1, '/boot')
+
+        self.assertFalse(self.fsm.bootable())
+
+        partition2 = self.fsm.add_partition(disk=disk, size=(1 * MB),
+                                            flag='boot')
+        fs2 = self.fsm.add_filesystem(partition2, 'ext4')
+        self.fsm.add_mount(fs2, '/')
+
+        self.assertTrue(self.fsm.bootable())
+
+    def test_can_install(self):
+        """can_install True if bootable and / in mountpoint mapping"""
+        self.fsm.probe()
+        all_disks = self.fsm.all_disks()
+        disk = random.choice(all_disks)
+        partition1 = self.fsm.add_partition(disk=disk, size=(1 * MB))
+        fs1 = self.fsm.add_filesystem(partition1, 'ext4')
+        self.fsm.add_mount(fs1, '/')
+
+        self.assertFalse(self.fsm.can_install())
+
+        partition2 = self.fsm.add_partition(disk=disk, size=(1 * MB),
+                                            flag='boot')
+        fs2 = self.fsm.add_filesystem(partition2, 'ext4')
+        self.fsm.add_mount(fs2, '/boot')
+
+        self.assertTrue(self.fsm.can_install())
+
+    def test_add_swap(self):
+        """verify add_swapfile returns True when swapfile is needed"""
+        self.fsm.probe()
+        all_disks = self.fsm.all_disks()
+        disk = random.choice(all_disks)
+        partition = self.fsm.add_partition(disk=disk, size=(1 * MB))
+        fs = self.fsm.add_filesystem(partition, 'ext4')
+        self.fsm.add_mount(fs, '/')
+        partition2 = self.fsm.add_partition(disk=disk, size=(1 * MB))
+        fs2 = self.fsm.add_filesystem(partition2, 'ext4')
+        self.fsm.add_mount(fs2, '/home')
+
+        self.assertTrue(self.fsm.add_swapfile())
+
+    def test_add_swap_false_with_btrfs(self):
+        """verify add_swapfile returns False when / is btrfs"""
+        self.fsm.probe()
+        all_disks = self.fsm.all_disks()
+        disk = random.choice(all_disks)
+        partition = self.fsm.add_partition(disk=disk, size=(1 * MB))
+        fs = self.fsm.add_filesystem(partition, 'btrfs')
+        self.fsm.add_mount(fs, '/')
+        self.assertFalse(self.fsm.add_swapfile())
+
+    def test_add_swap_false_with_swap(self):
+        """verify add_swapfile returns False when swap fs is present"""
+        self.fsm.probe()
+        all_disks = self.fsm.all_disks()
+        disk = random.choice(all_disks)
+        partition = self.fsm.add_partition(disk=disk, size=(1 * MB))
+        self.fsm.add_filesystem(partition, 'swap')
+
+        self.assertFalse(self.fsm.add_swapfile())
+
+    def test_render(self):
```

Comment:
I think this test is probably baking in too many implementation details. Like at the very least, compare the data structures rather than comparing the serialized yaml?

I think what we care about is that all the disks/partitions/etc objects are represented in the config, and that anything referred to by id is defined before the thing that refers to it. All else seems like detail to me, am I forgetting something?